WordPressAuthenticator/GoogleSignIn/NewGoogleAuthenticator.swift

```diff
@@ -0,0 +1,97 @@
+import AuthenticationServices
+
+public class NewGoogleAuthenticator: NSObject {
+
+    let clientId: GoogleClientId
+    let scheme: String
+    let audience: String
+    let contextProvider: ASWebAuthenticationPresentationContextProviding
+
+    let oauthTokenGetter: GoogleOAuthTokenGetting
+
+    public convenience init(
+        clientId: GoogleClientId,
+        scheme: String,
+        audience: String,
+        contextProvider: ASWebAuthenticationPresentationContextProviding,
+        urlSession: URLSession
+    ) {
+        self.init(
+            clientId: clientId,
+            scheme: scheme,
+            audience: audience,
+            contextProvider: contextProvider,
+            oautTokenGetter: GoogleOAuthTokenGetter(dataGetter: urlSession)
+        )
+    }
+
+    init(
+        clientId: GoogleClientId,
+        scheme: String,
+        audience: String,
+        contextProvider: ASWebAuthenticationPresentationContextProviding,
+        oautTokenGetter: GoogleOAuthTokenGetting
+    ) {
+        self.clientId = clientId
+        self.scheme = scheme
+        self.audience = audience
+        self.oauthTokenGetter = oautTokenGetter
+        self.contextProvider = contextProvider
+    }
+
+    /// Get the user's OAuth token from their Google account. This token can be used to authenticate with the WordPress backend.
+    public func getOAuthToken() async throws -> IDToken {
+        let pkce = ProofKeyForCodeExchange()
+        let url = try await getURL(clientId: clientId, scheme: scheme, pkce: pkce)
+        return try await requestOAuthToken(url: url, clientId: clientId, audience: audience, pkce: pkce)
+    }
+
+    func getURL(clientId: GoogleClientId, scheme: String, pkce: ProofKeyForCodeExchange) async throws -> URL {
+        let url = try URL.googleSignInAuthURL(clientId: clientId, pkce: pkce)
+        return try await withCheckedThrowingContinuation { continuation in
+            let session = ASWebAuthenticationSession(
+                url: url,
+                callbackURLScheme: scheme,
+                completionHandler: { result in
+                    continuation.resume(with: result)
+                }
+            )
+
+            session.presentationContextProvider = contextProvider
+            // FIXME: should this be configurable and/or defaulting to what?
+            session.prefersEphemeralWebBrowserSession = false
+
+            // FIXME: Need this to avoid TSAN error in demo app. Can we move the call elsewhere?
+            DispatchQueue.main.async {
+                session.start()
+            }
```

Comment:
Keen on ideas on this. But it could also be that it's okay to start the session on the main thread from here.

Reply:
I think the only way to get around this would be to write a callback-based version of this method that uses an assertion to ensure it's running on the main thread. Then from there, wrap _that_ in `withCheckedThrowingContinuation` and mark the parent method as `@MainThread` – I'm not 100% sure that'll work because I don't know if `withCheckedThrowingContinuation` guarantees that it runs its code on the parent Task context though.

Alternatively, you could look at a delegate-based approach which would let you make guarantees about which thread that call runs on, but it'd be more complex.

Reply:
I just tried this without the `DispatchQueue.main.async` [in WordPress](https://github.com/wordpress-mobile/WordPress-iOS/pull/20128) and got the same kind of error I got with the demo app

```
=================================================================
Main Thread Checker: UI API called on a background thread: -[UIViewController view]
PID: 25962, TID: 52550464, Thread name: (none), Queue name: com.apple.root.user-initiated-qos.cooperative, QoS: 25
Backtrace:
4   WordPress                           0x0000000105261db0 $s22WordPressAuthenticator24GoogleAuthViewControllerC18presentationAnchor3forSo8UIWindowCSo26ASWebAuthenticationSessionC_tF + 132
5   WordPress                           0x0000000105261f24 $s22WordPressAuthenticator24GoogleAuthViewControllerC18presentationAnchor3forSo8UIWindowCSo26ASWebAuthenticationSessionC_tFTo + 52
6   AuthenticationServices              0x000000011033a558 -[ASWebAuthenticationSession _startDryRun:] + 92
7   WordPress                           0x00000001052d8fc0 $s22WordPressAuthenticator09NewGoogleC0C6getURL8clientId6scheme4pkce10Foundation0G0VAA0e6ClientI0V_SSAA23ProofKeyForCodeExchangeVtYaKFyScCyAJs5Error_pGXEfU_ + 512
```

I'd say for the time being it should be safer to keep the dispatch call in here and document why we did so.

Reply:
Dropped a comment about this with ecd5888d